lib/Optimizer/GraphOptimizer/NodeSplitting.cpp

```diff
@@ -0,0 +1,1561 @@
+
+    // Set clone output types. The original node output types are not
+    // modified because the clone owns its output types.
+    for (const auto &outputIdxMap : outputIdxAndMaps) {
+      auto &outputRange = outputRanges[outputIdxMap.first];
+      auto outputType = Type::newShape(*node->getType(outputIdxMap.first),
+                                       outputRange.getSizes());
+      outputTypes.push_back(outputType);
+      clone->getNthResult(outputIdxMap.first)
+          .setTypeUnsafe(&outputTypes.back());
+    }
+
+    // Modify clone.
+    splitNodeModifier(node, clone, inputRanges, outputRanges);
+
+    // Verify clone. If the clone is invalid at this point this means there
+    // is a logic error in the splitting infrastructure (the input/output
+    // maps are not checked properly or the split node modifier is flawed)
+    // so we throw an error (not the same thing as returning false which is
+    // intended for signaling that the splitting infrastructure correctly
+    // identified an incorrect split configuration or the split configuration
+    // is not accepted by the user constraints).
+    if (!clone->verify()) {
+      // Dump some extra error context.
+      llvm::errs() << "Slice range description:\n";
+      for (unsigned idx = 0, e = clone->getNumInputs(); idx < e; ++idx) {
+        llvm::errs() << clone->getInputName(idx) << ": "
+                     << inputRanges[idx].toString() << "\n";
+      }
+      for (unsigned idx = 0, e = clone->getNumResults(); idx < e; ++idx) {
+        llvm::errs() << clone->getOutputName(idx).str() << ": "
+                     << outputRanges[idx].toString() << "\n";
+      }
+      llvm::errs() << "Node description:\n";
+      llvm::errs() << clone->toString() << "\n";
+      RETURN_ERR("Invalid node obtained during node splitting!");
+    }
+
+    // Early break.
+    if (!splitNodesCheck) {
+      break;
+    }
+  }
+
+  // Check split nodes against user constraint (if any).
+  if (splitConstraint) {
+    splitNodesCheck = splitNodesCheck && (*splitConstraint)(node, splitNodes);
+  }
+
+  // Explicitly destroy the temporary nodes.
+  for (auto *splitNode : splitNodes) {
+    Node::destroyNode(splitNode);
+  }
+
+  return splitNodesCheck;
+}
+
+///===---------------------------------------------------------------------===//
+///                            splitAndReplaceNode
+///===---------------------------------------------------------------------===//
+static Expected<std::vector<Node *>> splitAndReplaceNode(
+    Node *node, const SplitNodeOption *splitOption,
+    const SplitNodeConstraint *splitConstraint, dim_t splitOutputIdx,
+    const llvm::ArrayRef<OpIdxAndMap> &inputIdxAndMaps,
+    const llvm::ArrayRef<OpIdxAndMap> &outputIdxAndMaps = {},
+    const SplitNodeModifier &splitNodeModifier = SplitNodeModifierNop) {
+
+  // If the split output operand has no dimensions then return.
+  if (!node->getType(splitOutputIdx)->dims().size()) {
```

Comment:
May be use `empty` instead of `!size()`?

Reply:
Done.